frontend/public/components/modals/column-management-modal.tsx

```diff
@@ -47,55 +54,53 @@ const DataListRow: React.FC<DataListRowProps> = ({ column, onChange, disableUnch
 export const ColumnManagementModal: React.FC<ColumnManagementModalProps> = ({
   cancel,
   close,
-  columnManagementID,
+  columnManagementId,
   columnManagementType,
   selectedColumns,
 }) => {
   const dispatch = useDispatch();
-  const initialDefaultColumns = !_.isEmpty(selectedColumns)
-    ? selectedColumns.filter((column) => column.title.length > 0 && !column.additional)
-    : {};
-  const initialAdditionalColumns = !_.isEmpty(selectedColumns)
-    ? selectedColumns.filter((column) => column.additional)
-    : {};
-  const [defaultColumns, setDefaultColumns] = React.useState<any[]>(initialDefaultColumns);
-
-  const [additionalColumns, setAdditionalColumns] = React.useState<ColumnHeaderRow[]>(
-    initialAdditionalColumns,
+  const initialSelected = new Set<string>(
+    selectedColumns.filter((column) => column.visible && column.id).map((column) => column.id),
   );
+  const defaultColumns = !_.isEmpty(selectedColumns)
+    ? selectedColumns.filter((column) => column.id && !column.additional)
+    : [];
+  const additionalColumns = !_.isEmpty(selectedColumns)
+    ? selectedColumns.filter((column) => column.additional)
+    : [];
 
-  const updateColumns = (checked: boolean, id: string, columnArray: any): any => {
-    return columnArray.map((column) => {
-      if (!id || column.ID === id) {
-        return { ...column, visible: checked };
-      }
-      return column;
-    });
-  };
+  const [checkedColumns, setCheckedColumns] = React.useState<Set<string>>(new Set(initialSelected));
 
-  const defaultColumnChange = (checked: boolean, event: React.SyntheticEvent): void => {
-    setDefaultColumns(updateColumns(checked, event?.currentTarget?.id, defaultColumns));
-  };
-
-  const additionalColumnChange = (checked: boolean, event: React.SyntheticEvent): void => {
-    setAdditionalColumns(updateColumns(checked, event?.currentTarget?.id, additionalColumns));
+  const onColumnChange = (checked: boolean, event: React.SyntheticEvent): void => {
+    const updatedCheckedColumns = new Set(checkedColumns);
+    const selectedId = event?.currentTarget?.id;
+    updatedCheckedColumns.has(selectedId)
+      ? updatedCheckedColumns.delete(selectedId)
+      : updatedCheckedColumns.add(selectedId);
+    setCheckedColumns(updatedCheckedColumns);
   };
 
   const submit = (event): void => {
     event.preventDefault();
-    dispatch(setTableColumns(columnManagementID, [...defaultColumns, ...additionalColumns]));
+    dispatch(
+      setTableColumns(
+        columnManagementId,
+        selectedColumns.map((column) => {
+          return { ...column, visible: checkedColumns.has(column.id) };
```

Comment:
How much data about the column are we storing? I'm worried about upgrades here where things like the column headings change. Really we should only be storing a `Set` of selected columns IDs. Everything else about the column should be passed to the table or column managed modal by code.

I think this relates to the need for the `visible` property. The column definition itself should never change, and the visibility should be tracked separately in `localStorage` / Redux.

Reply:
@spadgett removed visible prop, only tracked the set of col id's in localstorage/redux. passed only the col props needed to col management modal.